pelican/__init__.py

```diff
@@ -162,46 +162,62 @@ def run(self):
                     pluralized_draft_pages,
                     time.time() - start_time))
 
-    def get_generator_classes(self):
-        generators = [ArticlesGenerator, PagesGenerator]
+    def _get_generator_classes(self):
+        discovered_generators = [
+            (ArticlesGenerator, None),
+            (PagesGenerator, None)
+        ]
 
-        if self.settings['TEMPLATE_PAGES']:
-            generators.append(TemplatePagesGenerator)
-        if self.settings['OUTPUT_SOURCES']:
-            generators.append(SourceFileGenerator)
+        if self.settings["TEMPLATE_PAGES"]:
+            discovered_generators.append((TemplatePagesGenerator, None))
 
-        for pair in signals.get_generators.send(self):
-            (funct, value) = pair
+        if self.settings["OUTPUT_SOURCES"]:
+            discovered_generators.append((SourceFileGenerator, None))
 
-            if not isinstance(value, Iterable):
-                value = (value, )
+        for receiver, values in signals.get_generators.send(self):
+            if not isinstance(values, Iterable):
+                values = (values,)
 
-            for v in value:
-                if isinstance(v, type):
-                    logger.debug('Found generator: %s', v)
-                    generators.append(v)
+                discovered_generators.extend(
+                    [(generator, receiver) for generator in values]
+                )
 
         # StaticGenerator must run last, so it can identify files that
         # were skipped by the other generators, and so static files can
         # have their output paths overridden by the {attach} link syntax.
-        generators.append(StaticGenerator)
+        discovered_generators.append((StaticGenerator, None))
+
+        generators = []
+
+        for generator, receiver in discovered_generators:
+            if receiver is None:
+                origin = "internal"
+            else:
+                origin = receiver.__module__
+
+            if not isinstance(generator, type):
+                logger.error("Generator %s (%s) cannot be loaded" % (generator, origin))
+                continue
+
+            logger.debug("Found generator: %s (%s)" % (generator.__name__, origin))
```

Comment:
Still formatting beforehand :). This will construct a string and *then* pass that to logging. What I meant was, pass the message template (with placeholders) and values to be replaced as arguments instead:

```python
logger.debug("Found generator: %s (%s)", generator.__name__, origin)
```

Reply:
Now, sometime I truly feel like an idiot :facepalm:

Reply:
OK I hope they are all fixed now